main/src/ca/uwaterloo/flix/language/phase/unification/set/SetUnification.scala

```diff
@@ -17,29 +17,284 @@
 package ca.uwaterloo.flix.language.phase.unification.set
 
 import ca.uwaterloo.flix.language.phase.unification.set.SetFormula.*
+import ca.uwaterloo.flix.util.Result
+
+import scala.collection.mutable
 
 object SetUnification {
 
   /**
     * The static parameters of set unification.
     *
-    * @param sveRecSizeThreshold if positive, [[sve]] will give up on formulas beyond this size.
+    * @param sizeThreshold if positive, [[solveWithInfo]] will give up before SVE if there are more
+    *                      equations than this
+    * @param permutationLimit if positive, the maximum number of permutations that
+    *                         [[svePermuations]] will try
+    * @param sveRecSizeThreshold if positive, [[sve]] will give up on formulas beyond this size
+    * @param svePermutationExitSize [[svePermuations]] will stop searching early if it finds a\
+    *                               substitution smaller than this
+    * @param debugging the amount of debugging information printed
+    */
+  final case class Options(
+                            sizeThreshold: Int = 1800,
+                            permutationLimit: Int = 10,
+                            sveRecSizeThreshold: Int = 1800 * 2,
+                            svePermutationExitSize: Int = 0,
+                            debugging: Options.Debugging = Options.Debugging.Nothing
+                          )
+
+  object Options {
+    sealed trait Debugging
+
+    object Debugging {
+      case object Nothing extends Debugging
+
+      case object All extends Debugging
+    }
+  }
+
+  /**
+    * Represents the running state of the solver.
+    *   - `eqs`: the remaining equations to solve
+    *   - `errEqs`: the remaining equations that have been found unsolvable
+    *   - `subst`: the current substitution, which has already been applied to `eqs`
+    *   - `lastProgressPhase`: the name of the last phase that was run and made progress
+    *   - `currentPhase`: the number of the last phase that was run
+    */
+  private class State(var eqs: List[Equation]) {
+    var errEqs: List[Equation] = Nil
+    var subst: SetSubstitution = SetSubstitution.empty
+    var lastProgressPhaseName: String = "Setup (Nothing)"
+    var lastProgressPhaseNumber: Int = 0
+    var currentPhaseNumber: Int = 0
+
+    def allEqs: List[Equation] = eqs ++ errEqs
+  }
+
+  /**
+    * Moves all non-pending equations into their own list for performance, not
+    * required for correctness.
+    * */
+  private def moveErrEquations(state: State): Unit = {
+    state.eqs = state.eqs.filter(eq => {
```

Comment:
A filter with a side-effect.. okay..

But why not just do two filters and assign to both fields:

state.eqs = ...
state.errEqs = ...

Much better

Reply:
(Performance cannot be a concern because there are so few equations)

Reply:
I read further and now I am coming back. I don't understand this optimization.

I don't think it can give any meaningful performance improvement (because it represents a type error).

It is not relevant for correctness, so it is just pure cruft. YAGNI.

Let us just keep one set of equations?

Reply:
I understand your point but at the same time I'm trying to build the solver bottom-up towards a fast full system solver not just customizing to the current need. But we can add it later.

I will say though that most of compilations have type errors in an IDE setting so I think its important to fail efficiently too. and leaving equations in `state.eqs` will cause them to be continuously traversed with substitutions during fixpoint loops

Reply:
But there are very few equations right? Large busts, but very few equations.